File: Year-2022/LeetCode June Challenge/MinCostClimbingStairs.js
let dp = new Array(1004);
var minCostClimbingStairsHelper = function(n, cost) {
    if(n == 0) return cost[0];
    if(n == 1) return cost[1];
    if(dp[n] != -1) return dp[n];
    dp[n] = cost[n] +  Math.min(minCostClimbingStairsHelper(n - 1, cost), minCostClimbingStairsHelper(n - 2, cost))
    return dp[n];
};
var minCostClimbingStairs = function(cost) {
    let n = cost.length;
    dp.fill(-1);
    let ans = Math.min(minCostClimbingStairsHelper(n - 1, cost), minCostClimbingStairsHelper(n - 2, cost));
    return ans;
};

let cost = [1,100,1,1,1,100,1,1,100,1];
console.log(minCostClimbingStairs(cost));

File: Year-2022/LeetCode June Challenge/MinCostClimbingStairs.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
private:
    int dp[1004];
    int minCostClimbingStairsHelper(int n, vector<int>& cost) {
        if(n == 0) return cost[0];
        if(n == 1) return cost[1];
        if(dp[n] != -1) return dp[n];
        dp[n] = cost[n] + min(minCostClimbingStairsHelper(n - 1, cost), minCostClimbingStairsHelper(n - 2, cost));
        return dp[n];
    }
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        memset(dp, -1, sizeof(dp));
        int ans = min(minCostClimbingStairsHelper(n - 1, cost), minCostClimbingStairsHelper(n - 2, cost));
        return ans;
    }
};
int main(){
    vector<int> cost = {1,100,1,1,1,100,1,1,100,1};
    Solution s;
    cout<<s.minCostClimbingStairs(cost)<<endl;
    return 0;
}